modules/autofill/autofill-impl/src/test/java/com/duckduckgo/autofill/AutofillJavascriptInterfaceTest.kt

```diff
@@ -108,6 +109,139 @@ class AutofillJavascriptInterfaceTest {
+                loginCredential(username = "username1", password = "bar"),
+                loginCredential(username = null, password = "bar"),
+                loginCredential(username = null, password = "bar"),
+                loginCredential(username = "username2", password = null),
+            )
+        )
+        initiateGetAutofillDataRequest()
+        assertCredentialsAvailable()
+        assertCredentialsContains({ it.username }, "username1", "username2")
+    }
+
+    @Test
+    fun whenRequestSpecifiesSubtypePasswordAndNoEntriesThenNoCredentialsCallbackInvoked() = runTest {
+        setupRequestForSubTypePassword()
+        initiateGetAutofillDataRequest()
+        whenever(autofillStore.getCredentials(any())).thenReturn(emptyList())
+        assertCredentialsUnavailable()
+    }
+
+    @Test
+    fun whenRequestSpecifiesSubtypePasswordAndNoEntriesWithAPasswordThenNoCredentialsCallbackInvoked() = runTest {
+        setupRequestForSubTypePassword()
+        whenever(autofillStore.getCredentials(any())).thenReturn(
+            listOf(
+                loginCredential(username = "foo", password = null),
+                loginCredential(username = "bar", password = null)
+            )
+        )
+        initiateGetAutofillDataRequest()
+        assertCredentialsUnavailable()
+    }
+
+    @Test
+    fun whenRequestSpecifiesSubtypePasswordAndSingleEntryWithAPasswordThenCredentialsAvailableCallbackInvoked() = runTest {
+        setupRequestForSubTypePassword()
+        whenever(autofillStore.getCredentials(any())).thenReturn(
+            listOf(
+                loginCredential(username = null, password = null),
+                loginCredential(username = "foobar", password = null),
+                loginCredential(username = "foo", password = "bar")
+            )
+        )
+        initiateGetAutofillDataRequest()
+        assertCredentialsAvailable()
+        assertCredentialsContains({ it.password }, "bar")
+    }
+
+    @Test
+    fun whenRequestSpecifiesSubtypePasswordAndMultipleEntriesWithAPasswordThenCredentialsAvailableCallbackInvoked() = runTest {
+        setupRequestForSubTypePassword()
+        whenever(autofillStore.getCredentials(any())).thenReturn(
+            listOf(
+                loginCredential(username = null, password = null),
+                loginCredential(username = "username2", password = null),
+                loginCredential(username = "username1", password = "password1"),
+                loginCredential(username = null, password = "password2"),
+                loginCredential(username = null, password = "password3"),
+
+            )
+        )
+        initiateGetAutofillDataRequest()
+        assertCredentialsAvailable()
+        assertCredentialsContains({ it.password }, "password1", "password2", "password3")
+    }
+
+    private fun assertCredentialsContains(property: (LoginCredentials) -> String?, vararg expected: String?) {
```

Comment:
I think this can be simplified to:
```
private fun assertCredentialsContains(property: (LoginCredentials) -> String?, vararg expected: String?) {
        val numberExpected = expected.size
        val numberMatched = TestCallback.credentials?.filter { expected.contains(property(it)) }?.count()
        assertEquals("Wrong number of matched properties. Expected $numberExpected but found $numberMatched", numberExpected, numberMatched)
    }
```

Reply:
Neat! good suggestion. done